src/rdkafka_msgset_reader.c

```diff
@@ -0,0 +1,1076 @@
+                                        NULL : Value.data);
+
+        /* Assign message timestamp.
+         * If message was in a compressed MessageSet and the outer/wrapper
+         * Message.Attribute had a LOG_APPEND_TIME set, use the
+         * outer timestamp */
+        if (msetr->msetr_outer.tstype == RD_KAFKA_TIMESTAMP_LOG_APPEND_TIME) {
+                rkm->rkm_timestamp = msetr->msetr_outer.timestamp;
+                rkm->rkm_tstype    = msetr->msetr_outer.tstype;
+
+        } else if (hdr.MagicByte >= 1 && hdr.Timestamp) {
+                rkm->rkm_timestamp = hdr.Timestamp;
+                if (hdr.Attributes & RD_KAFKA_MSG_ATTR_LOG_APPEND_TIME)
+                        rkm->rkm_tstype = RD_KAFKA_TIMESTAMP_LOG_APPEND_TIME;
+                else
+                        rkm->rkm_tstype = RD_KAFKA_TIMESTAMP_CREATE_TIME;
+        }
+
+        /* Enqueue message on temporary queue */
+        rd_kafka_q_enq(&msetr->msetr_rkq, rko);
+        msetr->msetr_msgcnt++;
+
+        return RD_KAFKA_RESP_ERR_NO_ERROR; /* Continue */
+
+ err_parse:
+        /* Count all parse errors as partial message errors. */
+        rd_atomic64_add(&msetr->msetr_rkb->rkb_c.rx_partial, 1);
+        return rkbuf->rkbuf_err;
+}
+
+
+
+/**
+ * @brief MessageSet reader for MsgVersion v0..1 (FetchRequest v0..3)
+ */
+static rd_kafka_resp_err_t
+rd_kafka_msgset_reader_v0_1 (rd_kafka_msgset_reader_t *msetr) {
+        rd_kafka_buf_t *rkbuf = msetr->msetr_rkbuf;
+
+        /* Read messages until MessageSet is exhausted */
+        while (rd_kafka_buf_read_remain(rkbuf) > 0) {
+                rd_kafka_resp_err_t err;
+                err = rd_kafka_msgset_reader_msg_v0_1(msetr);
+                if (unlikely(err))
+                        break;
+        }
+
+        return RD_KAFKA_RESP_ERR_NO_ERROR;
+}
+
+
+
+
+
+/**
+ * @brief Message parser for MsgVersion v2
+ */
+static rd_kafka_resp_err_t
+rd_kafka_msgset_reader_msg_v2 (rd_kafka_msgset_reader_t *msetr) {
+        rd_kafka_buf_t *rkbuf = msetr->msetr_rkbuf;
+        rd_kafka_toppar_t *rktp = msetr->msetr_rktp;
+        struct {
+                int64_t Length;
+                int8_t  MsgAttributes;
+                int64_t TimestampDelta;
+                int64_t OffsetDelta;
+                int64_t Offset;  /* Absolute offset */
+                rd_kafkap_bytes_t Key;
+                rd_kafkap_bytes_t Value;
+                int64_t HeaderCnt;
+        } hdr;
+        rd_kafka_op_t *rko;
+        rd_kafka_msg_t *rkm;
+        /* Only log decoding errors if protocol debugging enabled. */
+        int log_decode_errors = (rkbuf->rkbuf_rkb->rkb_rk->rk_conf.debug &
+                                 RD_KAFKA_DBG_PROTOCOL);
+        size_t message_end;
+
+        rd_kafka_buf_read_varint(rkbuf, &hdr.Length);
+        message_end = rd_slice_offset(&rkbuf->rkbuf_reader) + hdr.Length;
+        rd_kafka_buf_read_varint(rkbuf, &hdr.MsgAttributes);
```

Comment:
I will update the KIP to reflect this (if I have not already), but we moved the control flag to one the batch attributes. When handling messages on the replica, we found that we needed a way to know whether the batch contained transactional markers. With a message-level attribute, this would require decompressing the batch, so instead we moved the attribute to the batch so that _all_ messages contained in the batch are considered control messages.